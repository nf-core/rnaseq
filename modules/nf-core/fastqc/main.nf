process FASTQC {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/fastqc:0.12.1--hdfd78af_0' :
        'biocontainers/fastqc:0.12.1--hdfd78af_0' }"

    input:
    tuple val(meta), path(reads)

    output:
    tuple val(meta), path("*.html"), emit: html
    tuple val(meta), path("*.zip") , emit: zip
    path  "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    // Make list of old name and new name pairs to use for renaming in the bash while loop
    def old_new_pairs = reads instanceof Path || reads.size() == 1 ? [[ reads, "${prefix}.${reads.extension}" ]] : reads.withIndex().collect { entry, index -> [ entry, "${prefix}_${index + 1}.${entry.extension}" ] }
    def rename_to = old_new_pairs*.join(' ').join(' ')
    def renamed_files = old_new_pairs.collect{ old_name, new_name -> new_name }.join(' ')
    """
    printf "%s %s\\n" $rename_to | while read old_name new_name; do
        [ -f "\${new_name}" ] || ln -s \$old_name \$new_name
    done

    fastqc \\
        $args \\
        --threads $task.cpus \\
        $renamed_files

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fastqc: \$( fastqc --version | sed '/FastQC v/!d; s/.*v//' )
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.html
    touch ${prefix}.zip

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fastqc: \$( fastqc --version | sed '/FastQC v/!d; s/.*v//' )
    END_VERSIONS
    """
}
