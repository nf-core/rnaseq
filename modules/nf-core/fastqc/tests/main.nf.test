nextflow_process {

    name "Test Process FASTQC"
    script "../main.nf"
    process "FASTQC"

    tag "modules"
    tag "modules_nfcore"
    tag "fastqc"

    test("sarscov2 single-end [fastq]") {

        when {
            process {
                """
                input[0] = Channel.of([
                    [ id: 'test', single_end:true ],
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true) ]
                ])
                """
            }
        }

        then {
            assertAll (
            { assert process.success },

            // NOTE The report contains the date inside it, which means that the md5sum is stable per day, but not longer than that. So you can't md5sum it.
            // looks like this: <div id="header_filename">Mon 2 Oct 2023<br/>test.gz</div>
            // https://github.com/nf-core/modules/pull/3903#issuecomment-1743620039

            { assert process.out.html[0][1] ==~ ".*/test_fastqc.html" },
            { assert process.out.zip[0][1] ==~ ".*/test_fastqc.zip" },
            { assert path(process.out.html[0][1]).text.contains("<tr><td>File type</td><td>Conventional base calls</td></tr>") },

            { assert snapshot(process.out.versions).match("fastqc_versions_single") }
            )
        }
    }

    test("sarscov2 paired-end [fastq]") {

        when {
            process {
                """
                input[0] = Channel.of([
                    [id: 'test', single_end: false], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true) ]
                ])
                """
            }
        }

        then {
            assertAll (
            { assert process.success },

            { assert process.out.html[0][1][0] ==~ ".*/test_1_fastqc.html" },
            { assert process.out.html[0][1][1] ==~ ".*/test_2_fastqc.html" },
            { assert process.out.zip[0][1][0] ==~ ".*/test_1_fastqc.zip" },
            { assert process.out.zip[0][1][1] ==~ ".*/test_2_fastqc.zip" },
            { assert path(process.out.html[0][1][0]).text.contains("<tr><td>File type</td><td>Conventional base calls</td></tr>") },
            { assert path(process.out.html[0][1][1]).text.contains("<tr><td>File type</td><td>Conventional base calls</td></tr>") },

            { assert snapshot(process.out.versions).match("fastqc_versions_paired") }
            )
        }
    }

    test("sarscov2 interleaved [fastq]") {

        when {
            process {
                """
                input[0] = Channel.of([
                    [id: 'test', single_end: false], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_interleaved.fastq.gz', checkIfExists: true)
                ])
            """
            }
        }

        then {
            assertAll (
            { assert process.success },

            { assert process.out.html[0][1] ==~ ".*/test_fastqc.html" },
            { assert process.out.zip[0][1] ==~ ".*/test_fastqc.zip" },
            { assert path(process.out.html[0][1]).text.contains("<tr><td>File type</td><td>Conventional base calls</td></tr>") },

            { assert snapshot(process.out.versions).match("fastqc_versions_interleaved") }
            )
        }
    }

    test("sarscov2 paired-end [bam]") {

        when {
            process {
                """
                input[0] = Channel.of([
                    [id: 'test', single_end: false], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll (
            { assert process.success },

            { assert process.out.html[0][1] ==~ ".*/test_fastqc.html" },
            { assert process.out.zip[0][1] ==~ ".*/test_fastqc.zip" },
            { assert path(process.out.html[0][1]).text.contains("<tr><td>File type</td><td>Conventional base calls</td></tr>") },

            { assert snapshot(process.out.versions).match("fastqc_versions_bam") }
            )
        }
    }

    test("sarscov2 multiple [fastq]") {

        when {
            process {
                """
                input[0] = Channel.of([
                    [id: 'test', single_end: false], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test2_1.fastq.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test2_2.fastq.gz', checkIfExists: true) ]
                ])
                """
            }
        }

        then {
            assertAll (
            { assert process.success },

            { assert process.out.html[0][1][0] ==~ ".*/test_1_fastqc.html" },
            { assert process.out.html[0][1][1] ==~ ".*/test_2_fastqc.html" },
            { assert process.out.html[0][1][2] ==~ ".*/test_3_fastqc.html" },
            { assert process.out.html[0][1][3] ==~ ".*/test_4_fastqc.html" },
            { assert process.out.zip[0][1][0] ==~ ".*/test_1_fastqc.zip" },
            { assert process.out.zip[0][1][1] ==~ ".*/test_2_fastqc.zip" },
            { assert process.out.zip[0][1][2] ==~ ".*/test_3_fastqc.zip" },
            { assert process.out.zip[0][1][3] ==~ ".*/test_4_fastqc.zip" },
            { assert path(process.out.html[0][1][0]).text.contains("<tr><td>File type</td><td>Conventional base calls</td></tr>") },
            { assert path(process.out.html[0][1][1]).text.contains("<tr><td>File type</td><td>Conventional base calls</td></tr>") },
            { assert path(process.out.html[0][1][2]).text.contains("<tr><td>File type</td><td>Conventional base calls</td></tr>") },
            { assert path(process.out.html[0][1][3]).text.contains("<tr><td>File type</td><td>Conventional base calls</td></tr>") },

            { assert snapshot(process.out.versions).match("fastqc_versions_multiple") }
            )
        }
    }

    test("sarscov2 custom_prefix") {

        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'mysample', single_end:true ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll (
            { assert process.success },

            { assert process.out.html[0][1] ==~ ".*/mysample_fastqc.html" },
            { assert process.out.zip[0][1] ==~ ".*/mysample_fastqc.zip" },
            { assert path(process.out.html[0][1]).text.contains("<tr><td>File type</td><td>Conventional base calls</td></tr>") },

            { assert snapshot(process.out.versions).match("fastqc_versions_custom_prefix") }
            )
        }
    }

    test("sarscov2 single-end [fastq] - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = Channel.of([
                    [ id: 'test', single_end:true ],
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true) ]
                ])
                """
            }
        }

        then {
            assertAll (
            { assert process.success },
            { assert snapshot(process.out.html.collect { file(it[1]).getName() } +
                                process.out.zip.collect { file(it[1]).getName() } +
                                process.out.versions ).match("fastqc_stub") }
            )
        }
    }

}
