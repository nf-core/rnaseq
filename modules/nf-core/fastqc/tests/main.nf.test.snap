{
    "fastqc_versions_interleaved": {
        "content": [
            [
                "versions.yml:md5,e1cc25ca8af856014824abd842e93978"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-31T17:40:07.293713"
    },
    "fastqc_stub": {
        "content": [
            [
                "test.html",
                "test.zip",
                "versions.yml:md5,e1cc25ca8af856014824abd842e93978"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-31T17:31:01.425198"
    },
    "fastqc_versions_multiple": {
        "content": [
            [
                "versions.yml:md5,e1cc25ca8af856014824abd842e93978"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-31T17:40:55.797907"
    },
    "fastqc_versions_bam": {
        "content": [
            [
                "versions.yml:md5,e1cc25ca8af856014824abd842e93978"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-31T17:40:26.795862"
    },
    "fastqc_versions_single": {
        "content": [
            [
                "versions.yml:md5,e1cc25ca8af856014824abd842e93978"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-31T17:39:27.043675"
    },
    "fastqc_versions_paired": {
        "content": [
            [
                "versions.yml:md5,e1cc25ca8af856014824abd842e93978"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-31T17:39:47.584191"
    },
    "fastqc_versions_custom_prefix": {
        "content": [
            [
                "versions.yml:md5,e1cc25ca8af856014824abd842e93978"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-31T17:41:14.576531"
    }
}